ocaml/typing/typecore.ml

```diff
@@ -2795,7 +2800,7 @@ and type_pat_aux
   | Ppat_constraint(sp_constrained, sty) ->
       (* Pretend separate = true *)
       let cty, ty, expected_ty' =
-        let type_modes = alloc_mode_from_pat_attrs sp in
+        let type_modes, sty = alloc_mode_from_typ_attrs sty in
```

Comment:
I fear that putting the attributes on the type is incompatible with a small detail of #2335: in that PR, we start tracking the mode in `let local_ x : int -> int = ...` on the type in "value binding elaboration" in the type-checker. So I'd like to take a look at this after you resolve conflicts and before we merge into `main`.

Reply:
Yeah, not only they are compatible, both changes are driven by the same idea: the said modes is only for the interpretation of the type in `Pexp_constraint` or `Ppat_constraint`, so the modes should be in `ptyp_attributes` instead of `pexp_attributes` or `ppat_attributes`.